Mathlib/Order/CompletePartialOrder.lean

```diff
@@ -0,0 +1,243 @@
+  set : Set α
+  directed : DirectedOn (. ≤ .) set
+
+variable [SemilatticeSup α]
+
+/--
+Every subset of a join-semilattice generates a directed set
+-/
+def Set.ToDirectedSet (s : Set α) : DirectedSet α := {
+  set := { a | ∃ F : Finset α, ∃ H : F.Nonempty, ↑F ⊆ s ∧  a = F.sup' H id   },
+  directed := by classical
+    intros a ha b hb
+    simp at ha hb
+    obtain ⟨Fa,hFa⟩ := ha
+    obtain ⟨Fb,hFb⟩ := hb
+    use a⊔b
+    constructor
+    · simp
+      use (Fa ⊔ Fb)
+      simp
+      simp at hFa
+      constructor
+      · constructor
+        · exact hFa.1
+        · exact hFb.1
+      · obtain ⟨hnFa,ha⟩ := hFa.2
+        obtain ⟨hnFb,hb⟩ := hFb.2
+        use (Finset.Nonempty.inl hnFa)
+        rw [le_antisymm_iff]
+        constructor
+        · simp
+          constructor
+          · rw [hFa.2.2]
+            apply Finset.sup'_mono
+            exact Finset.subset_union_left Fa Fb
+          · rw [hFb.2.2]
+            apply Finset.sup'_mono
+            exact Finset.subset_union_right Fa Fb
+        · simp
+          intros c hc
+          cases' hc with h₁ h₂
+          · apply le_sup_of_le_left
+            rw [ha]
+            exact Finset.le_sup'_of_le _ h₁ (Eq.le rfl)
+          · apply le_sup_of_le_right
+            rw [hb]
+            exact Finset.le_sup'_of_le _ h₂ (Eq.le rfl)
+    · constructor
+      · exact le_sup_left
+      · exact le_sup_right
+}
+
+lemma Set_subseteq_DirectedSet {s : Set α} :
+    s ⊆ ↑(Set.ToDirectedSet s).set := by
+  intro a ha
+  rw [Set.ToDirectedSet]
+  simp only [id_eq, exists_and_left, Set.mem_setOf_eq]
+  use ({a} : Finset α)
+  constructor
+  · exact Iff.mpr Finset.singleton_subset_set_iff ha
+  · use (Finset.singleton_nonempty a)
+    rfl
+
+lemma Set_DirectedSet_upperBounds {s : Set α} :
+    upperBounds (Set.ToDirectedSet s).set = upperBounds s := by
+  rw [subset_antisymm_iff]
+  constructor
+  · exact upperBounds_mono_set Set_subseteq_DirectedSet
+  · intro u hu
+    rw [mem_upperBounds]
+    intro b hb
+    obtain ⟨Fb,⟨H,hFb⟩⟩ := hb
+    rw [hFb.2, Finset.sup'_le_iff]
+    intro c hc
+    exact hu (hFb.1 hc)
+
+lemma Set_DirectedSet_LUB [SemilatticeSup α] {s : Set α} {u : α} : IsLUB s u ↔
+    IsLUB (Set.ToDirectedSet s).set u := by
+  constructor
+  · intro hsu
+    constructor
+    · rw [Set_DirectedSet_upperBounds, mem_upperBounds]
+      exact hsu.1
+    · rw [mem_lowerBounds]
+      intro b hb
+      rw [isLUB_le_iff hsu]
+      exact upperBounds_mono_set Set_subseteq_DirectedSet hb
+  · intro h
+    constructor
+    · rw [← Set_DirectedSet_upperBounds]
+      exact Set.mem_of_mem_inter_left h
+    · intro v hv
+      rw [← Set_DirectedSet_upperBounds] at hv
+      exact Iff.mpr (isLUB_le_iff h) hv
+
+/--
+A join semi-lattice where every directed subset has a least upper bound is automatically complete
+-/
+def SemilatticeSup.toCompleteSemilatticeSup (dSup : DirectedSet α → α)
+    (h : ∀ (d : DirectedSet α), IsLUB d.set (dSup d)) : CompleteSemilatticeSup α where
```

Comment:
and here you could do
```suggestion
def SemilatticeSup.toCompleteSemilatticeSup (dSup : Set α → α)
    (h : ∀ d, DirectedOn (. ≤ .) d → IsLUB d (dSup d)) : CompleteSemilatticeSup α where
```

Reply:
@YaelDillies not sure I follow this one... `dSup` is only defined on sets which are directed.

Reply:
I guess, if it's a `SemilatticeSup` a function `f` on the directed sets can always be extended to all sets by composition with `directedClosure`.